check whether the function sig can be simplify

miss push the commit to the `main`, need hotfix it.

```
$ cargo clippy
    Checking fsrs v3.0.0 (/Users/jarrettye/Codes/open-spaced-repetition/fsrs-rs)
error[E0782]: expected a type, found a trait
  --> src/optimal_retention.rs:42:37
   |
42 | pub struct PostSchedulingFn(pub Arc<Fn(f32, f32, usize, &[usize], &mut StdRng) -> f32>);
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: you can add the `dyn` keyword if you want a trait object
   |
42 | pub struct PostSchedulingFn(pub Arc<dyn Fn(f32, f32, usize, &[usize], &mut StdRng) -> f32>);
   |                                     +++

error[E0782]: expected a type, found a trait
  --> src/optimal_retention.rs:59:37
   |
59 | pub struct ReviewPriorityFn(pub Arc<Fn(&Card) -> i32>);
   |                                     ^^^^^^^^^^^^^^^^
   |
help: you can add the `dyn` keyword if you want a trait object
   |
59 | pub struct ReviewPriorityFn(pub Arc<dyn Fn(&Card) -> i32>);
   |                                     +++

For more information about this error, try `rustc --explain E0782`.
error: could not compile `fsrs` (lib) due to 2 previous errors
```

--- a/src/optimal_retention.rs
+++ b/src/optimal_retention.rs
@@ -39,7 +39,9 @@ const R_MAX: f32 = 0.95;
 /// Function type for post scheduling operations that takes interval, maximum interval,
 /// current day index, due counts per day, and a random number generator,
 /// and returns a new interval.
-pub struct PostSchedulingFn(pub Arc<Fn(f32, f32, usize, &[usize], &mut StdRng) -> f32>);
+pub struct PostSchedulingFn(
+    pub Arc<dyn Fn(f32, f32, usize, &[usize], &mut StdRng) -> f32 + Sync + Send>,
+);
 
 impl PartialEq for PostSchedulingFn {
     fn eq(&self, _: &Self) -> bool {
@@ -56,7 +58,7 @@ impl std::fmt::Debug for PostSchedulingFn {
 /// Function type for review priority calculation that takes a card reference
 /// and returns a priority value (lower value means higher priority)
 #[derive(Clone)]
-pub struct ReviewPriorityFn(pub Arc<Fn(&Card) -> i32>);
+pub struct ReviewPriorityFn(pub Arc<dyn Fn(&Card) -> i32 + Sync + Send>);
 
 impl PartialEq for ReviewPriorityFn {
     fn eq(&self, _: &Self) -> bool {
@@ -414,12 +416,12 @@ pub fn simulate(
             .round()
             .clamp(1.0, config.max_ivl);
 
-        if let Some(PostSchedulingFn(ref cb)) = config.post_scheduling_fn {
+        if let Some(PostSchedulingFn(cb)) = &config.post_scheduling_fn {
             ivl = cb(
                 ivl,
                 config.max_ivl,
                 day_index,
-                due_cnt_per_day.clone(),
+                &due_cnt_per_day.clone(),
                 &mut rng,
             );
         }

--- a/src/optimal_retention.rs
+++ b/src/optimal_retention.rs
@@ -418,13 +418,7 @@ pub fn simulate(
             .clamp(1.0, config.max_ivl);
 
         if let Some(PostSchedulingFn(cb)) = &config.post_scheduling_fn {
-            ivl = cb(
-                ivl,
-                config.max_ivl,
-                day_index,
-                &due_cnt_per_day,
-                &mut rng,
-            );
+            ivl = cb(ivl, config.max_ivl, day_index, &due_cnt_per_day, &mut rng);
         }
 
         card.last_date = day_index as f32;

--- a/src/optimal_retention.rs
+++ b/src/optimal_retention.rs
@@ -421,7 +421,7 @@ pub fn simulate(
                 ivl,
                 config.max_ivl,
                 day_index,
-                &due_cnt_per_day.clone(),
+                &due_cnt_per_day,
                 &mut rng,
             );
         }

--- a/src/optimal_retention.rs
+++ b/src/optimal_retention.rs
@@ -39,6 +39,7 @@ const R_MAX: f32 = 0.95;
 /// Function type for post scheduling operations that takes interval, maximum interval,
 /// current day index, due counts per day, and a random number generator,
 /// and returns a new interval.
+#[allow(clippy::type_complexity)]
 pub struct PostSchedulingFn(
     pub Arc<dyn Fn(f32, f32, usize, &[usize], &mut StdRng) -> f32 + Sync + Send>,
 );

